### ISSUE TITLE
Contributing Golang and Rust implementations

### ISSUE THREAD

**Issue by rw** [zarr-developers/zarr#437]
Hi there!

Via @ryan-williams, I'm posting here because I'm interested in making two ports of Zarr: to native Golang and to native Rust. (To me, "native" means "no FFI to an existing Zarr library".)

For background:

I wrote and maintain the official Golang, Python, Rust ports of [Google's FlatBuffers serialization library](https://github.com/google/flatbuffers/graphs/contributors). I received a Google Open Source Contributors Award for my volunteer efforts on FlatBuffers.

Here are some relevant links:
+ The port of FlatBuffers to Rust: https://github.com/google/flatbuffers/pull/4898
+ The port of FlatBuffers to Golang: https://github.com/google/flatbuffers/pull/36
+ The port of FlatBuffers to Python: https://github.com/google/flatbuffers/pull/112
+ The talk  I gave at Rust Bay Area at Mozilla HQ, titled [High-Performance Zero-copy Serialization in Native Rust with FlatBuffers](
https://www.youtube.com/watch?v=YsiQDX20lXI).

All of that is to say, that I have a background in writing high-performance serialization code in open-source projects.

So, is there a need from the community for Golang and/or Rust ports? I'm really interested in getting involved, because I like both your technical solutions, as well as the community-friendly group dynamics that I've seen.

Best,
Robert

**Comment by martindurant** [comment]
I am total novice with Rust, but have been looking for a project to teach myself on. Since I know something about the python implementation (although my main experience in this are is with pythonic file-systems), this might be the perfect entry point. 

As for need, I simply don't know - you would look for other science or other big-array workflows in the Rust world. Where some have been speaking of C/C++ implementations of the zarr spec, there is certainly an argument to trying to get the same performance, but with more modern code, via Rust.

**Comment by rw** [comment]
> I am total novice with Rust, but have been looking for a project to teach myself on.

@martindurant Great! I'm thinking I can use my Rust serialization experience to start the project, then would you be interested in helping to maintain it?

**Comment by constantinpape** [comment]
There is a rust implementation for n5 already:
https://github.com/aschampion/rust-n5

It should be easy to port this to zarr, given how close the specs are.

**Comment by jakirkham** [comment]
Thanks @constantinpape. Let’s see if @aschampion has thoughts. 🙂

**Comment by martindurant** [comment]
I don't want to promise too much, given how much I (don't) know...

> There is a rust implementation for n5 already

There is a chance of merging the libraries or spec eventually, right? (z5, whatever)

**Comment by rw** [comment]
> There is a rust implementation for n5 already:
> https://github.com/aschampion/rust-n5
> 
> It should be easy to port this to zarr, given how close the specs are.

@constantinpape It's not clear that that that project is written with performance in mind. Reading the code, I see multiple heap allocations throughout the code, in what I understand to be tight loops.

**Comment by constantinpape** [comment]
@martindurant  
> There is a chance of merging the libraries or spec eventually, right? (z5, whatever)

Yes, hopefully zarr spec v3 will merge zarr and n5. (I couldn't make the call today so I am not quite up-to-date on the progress regarding this).

@rw  
> It's not clear to me that that project is written with performance in mind. Reading the code, I see multiple heap allocations throughout the code, in what I understand to be tight loops.

That's probably something @aschampion could comment more on.

**Comment by constantinpape** [comment]
cc @j6k4m8 who might be interested in a Go implementation

**Comment by aschampion** [comment]
> @constantinpape I'm not an expert yet in Zarr nor N5, but after reading that code base, I think there are a lot of opportunities to improve its usage of the heap. The way it is written leads me to believe that it is not designed to minimize heap allocations.

There are opportunities for optimization. As is it's at least a bit faster than the Java reference implementation, which it started as a rather direct translation of

![java-rust-n5](https://user-images.githubusercontent.com/460206/57445310-f4fe5b00-7249-11e9-9b5d-f97e1357ba43.png)

I had a branch which reduced some allocations with `SmallVec`, etc., but since in practice my perf is usually IO bound the difference was marginal, and having fewer dependencies was critical to get wasm compilation working then. Because of improvement in the rust toolchain and ecosystem in the last years, that's less of a concern now. With modern rust one could easily get rid of the allocs for compression dispatch, etc.

The kludge for composing block reads/writes from ndarrays is a mess of needless allocations, but that's effectively an outer loop around block IO.

I don't have an immediate need or time to develop zarr support myself, but would happily accept PRs, including major restructuring, so long as the downstream [wasm](https://github.com/aschampion/n5-wasm) and [conda-less pip installable python](https://github.com/pattonw/rust-pyn5) packages are still possible. But I also understand if you'd rather to start from scratch.

**Comment by rw** [comment]
@alimanfoo I'm happy to start with the V3 spec, and give feedback as I go. Would you be able to fill in the TODOs in the v3 spec with your best guesses, so that I have a concrete description upon which to work? Or, perhaps, provide links to the relevant parts of the v2 spec to fill in the gaps?

**Comment by alimanfoo** [comment]
>
> @alimanfoo <https://github.com/alimanfoo> I'm happy to start with the V3
> spec, and give feedback as I go. Would you be able to fill in the TODOs in
> the v3 spec with your best guesses, so that I have a concrete description
> upon which to work?
>
Sure, working on it, hoping to have a complete sketch by the end of the
week. Please bear in mind it will be just a straw man for the moment, will
need some time for others to digest and comment. Follow along here for
progress and discussion:
https://github.com/zarr-developers/zarr-specs/pull/16

**Comment by rw** [comment]
@alimanfoo Hmm, should I just start with v2? I can still offer feedback on the v3 spec.

**Comment by alimanfoo** [comment]
> @alimanfoo Hmm, should I just start with v2? I can still offer feedback on the v3 spec.

That would be fine I'm sure, lots of value in having something that would interoperate with current Zarr implementations.
